### ISSUE TITLE
How Detect module works?

### ISSUE THREAD

**Issue by iumyx2612** [ultralytics/yolov5#6928]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Sorry because this is gonna be a really long question and this is gonna be the first part
You can construct [Detect ](https://github.com/ultralytics/yolov5/blob/master/models/yolo.py#L33)module from config file with `[[17, 20, 23], 1, Detect, [nc, anchors]]` 
and anchors are also modified in the config file:
```
anchors:
  - [10,13, 16,30, 33,23]  # P3/8
  - [30,61, 62,45, 59,119]  # P4/16
  - [116,90, 156,198, 373,326]  # P5/32
```
Can you explain these parameters?

**Comment by spacewalk01** [comment]
I think that is width-height sizes of anchor boxes you want to use for object detection.

**Comment by iumyx2612** [comment]
> I think that is width-height sizes of anchor boxes you want to use for object detection.

So on every single location of output feature map from P3, create 3 anchor boxes with size (10, 13), (16, 30) and (33, 23) respectively?

**Comment by jedi007** [comment]
These parameters are yolov3 passed down。
They are calculated by K-means on the coco data set。

**Comment by iumyx2612** [comment]
> These parameters are yolov3 passed down。 They are calculated by K-means on the coco data set。

So what do those numbers mean? Size of anchors in pixels?

**Comment by jedi007** [comment]
Yes, "- [10,13, 16,30, 33,23]  # P3/8" means 3 anchor boxes in P3,their width and height are [10 pixels, 13 pixels],[16 pixels, 30 pixels],[33 pixels, 23 pixels]

**Comment by iumyx2612** [comment]
> Yes, "- [10,13, 16,30, 33,23] # P3/8" means 3 anchor boxes in P3,their width and height are [10 pixels, 13 pixels],[16 pixels, 30 pixels],[33 pixels, 23 pixels]

But the feature map from layer 23 is the tiniest with its dimension being (h/32, w/32). Somehow the anchors have the largest values: `[116,90, 156,198, 373,326]  # P5/32`. Isn't this quite of weird?

**Comment by glenn-jocher** [comment]
@iumyx2612 anchor units are expressed in input image pixels, i.e. they should match the size of objects in 640-size images if you use default training settings.

**Comment by iumyx2612** [comment]
> @iumyx2612 anchor units are expressed in input image pixels, i.e. they should match the size of objects in 640-size images if you use default training settings.

Thank you, I think I'll learn some more about anchors then I'll continue asking later

**Comment by iumyx2612** [comment]
@glenn-jocher 
So in training, Detect module will output a tensor of shape (batch_size, 3, feature_map_height, feature_map_width, num_class + 5)  
To calculate the loss, it will then feed into [ComputeLoss](https://github.com/ultralytics/yolov5/blob/master/utils/loss.py#L117) along with targets from [LoadImagesAndLabels](https://github.com/ultralytics/yolov5/blob/master/utils/datasets.py#L621). This target is (0, class, x, y, w, h). **First question: Why do we need 0 here?**
So from YOLOv2, YOLO has import anchors in its model. To calculate the loss, we need to instantiate bunch of anchors on the input images, find which anchors responsible detecting object, called positive labels, by calculating IoU of (anchors, gt boxes). Do we build these anchors in [build_targets](https://github.com/ultralytics/yolov5/blob/master/utils/loss.py#L169)? I actually can't understand how do you build those anchors, match them with gt boxes and calculate loss. **Can you walk me through the [build_targets](https://github.com/ultralytics/yolov5/blob/master/utils/loss.py#L169) function** . I'm having a really hard time following what happens in it

**Comment by glenn-jocher** [comment]
@iumyx2612 0 is the image index in the batch.

build_targets() builds a list of label-prediction combinations to create losses for.
